File: system/store/record.c
#include "yai/store/record.h"

#include <stddef.h>
#include <string.h>

void yai_copy_string(char *dst, size_t dst_size, const char *src);

yai_status_t yai_store_record_init(yai_store_record_t *record,
                                   const char *record_id,
                                   const yai_case_ref_t *case_ref,
                                   yai_record_kind_t record_kind,
                                   const yai_subject_ref_t *subject_ref,
                                   const yai_id_t *attempt_id,
                                   const yai_id_t *decision_id,
                                   const yai_id_t *receipt_id,
                                   const char *summary) {
    if (record == 0 || record_id == 0 || record_id[0] == '\0' ||
        yai_case_ref_is_empty(case_ref)) {
        return YAI_ERR_INVALID;
    }

    yai_id_set(&record->record_id, record_id);
    record->case_ref = *case_ref;
    record->record_kind = record_kind;

    if (subject_ref != 0) {
        record->subject_ref = *subject_ref;
    } else {
        (void)yai_subject_ref_init(&record->subject_ref, "subject:none", "external_system", "none");
    }

    if (attempt_id != 0) {
        record->attempt_id = *attempt_id;
    } else {
        yai_id_set(&record->attempt_id, "");
    }

    if (decision_id != 0) {
        record->decision_id = *decision_id;
    } else {
        yai_id_set(&record->decision_id, "");
    }

    if (receipt_id != 0) {
        record->receipt_id = *receipt_id;
    } else {
        yai_id_set(&record->receipt_id, "");
    }

    yai_copy_string(record->summary, sizeof(record->summary), summary);
    return YAI_OK;
}

const char *yai_record_kind_string(yai_record_kind_t kind) {
    switch (kind) {
    case YAI_RECORD_CASE:
        return "case";
    case YAI_RECORD_SUBJECT_BINDING:
        return "subject_binding";
    case YAI_RECORD_ATTEMPT:
        return "attempt";
    case YAI_RECORD_DECISION:
        return "decision";
    case YAI_RECORD_RECEIPT:
        return "receipt";
    case YAI_RECORD_PROJECTION:
        return "projection";
    case YAI_RECORD_POLICY_RULE:
        return "policy_rule";
    case YAI_RECORD_GATE_RESULT:
        return "gate_result";
    case YAI_RECORD_DECISION_BASIS:
        return "decision_basis";
    case YAI_RECORD_OBLIGATION:
        return "obligation";
    case YAI_RECORD_RECEIPT_REQUIREMENT:
        return "receipt_requirement";
    case YAI_RECORD_CARRIER_REQUEST:
        return "carrier_request";
    case YAI_RECORD_EFFECT_RECEIPT:
        return "effect_receipt";
    case YAI_RECORD_FILESYSTEM_RECEIPT:
        return "filesystem_receipt";
    case YAI_RECORD_SUBJECT_STATE:
        return "subject_state";
    case YAI_RECORD_GRAPH_EDGE:
        return "graph_edge";
    case YAI_RECORD_RECONSTRUCTION:
        return "reconstruction";
    case YAI_RECORD_MEMORY_CANDIDATE:
        return "memory_candidate";
    case YAI_RECORD_DIVERGENCE:
        return "divergence";
    case YAI_RECORD_RECONCILIATION:
        return "reconciliation";
    case YAI_RECORD_PROJECTION_REQUEST:
        return "projection_request";
    case YAI_RECORD_PROJECTION_RESULT:
        return "projection_result";
    case YAI_RECORD_QUERY_RESULT:
        return "query_result";
    case YAI_RECORD_PROJECTION_RULE:
        return "projection_rule";
    case YAI_RECORD_AUTHORITY_SCOPE:
        return "authority_scope";
    case YAI_RECORD_MODEL_INTERPRETATION:
        return "model_interpretation";
    case YAI_RECORD_CASE_DOMAIN:
        return "case_domain";
    case YAI_RECORD_CASE_ATTACHMENT:
        return "case_attachment";
    case YAI_RECORD_CASE_BINDING:
        return "case_binding";
    case YAI_RECORD_INTERACTION_THREAD:
        return "interaction_thread";
    case YAI_RECORD_INTERACTION_TURN:
        return "interaction_turn";
    case YAI_RECORD_PARTICIPANT_VIEW_FRAME:
        return "participant_view_frame";
    default:
        return "unknown";
    }
}

yai_status_t yai_record_kind_from_string(const char *value,
                                         yai_record_kind_t *kind) {
    if (value == 0 || kind == 0) {
        return YAI_ERR_INVALID;
    }

    if (strcmp(value, "case") == 0) {
        *kind = YAI_RECORD_CASE;
    } else if (strcmp(value, "subject_binding") == 0) {
        *kind = YAI_RECORD_SUBJECT_BINDING;
    } else if (strcmp(value, "attempt") == 0) {
        *kind = YAI_RECORD_ATTEMPT;
    } else if (strcmp(value, "decision") == 0) {
        *kind = YAI_RECORD_DECISION;
    } else if (strcmp(value, "receipt") == 0) {
        *kind = YAI_RECORD_RECEIPT;
    } else if (strcmp(value, "projection") == 0) {
        *kind = YAI_RECORD_PROJECTION;
    } else if (strcmp(value, "policy_rule") == 0) {
        *kind = YAI_RECORD_POLICY_RULE;
    } else if (strcmp(value, "gate_result") == 0) {
        *kind = YAI_RECORD_GATE_RESULT;
    } else if (strcmp(value, "decision_basis") == 0) {
        *kind = YAI_RECORD_DECISION_BASIS;
    } else if (strcmp(value, "obligation") == 0) {
        *kind = YAI_RECORD_OBLIGATION;
    } else if (strcmp(value, "receipt_requirement") == 0) {
        *kind = YAI_RECORD_RECEIPT_REQUIREMENT;
    } else if (strcmp(value, "carrier_request") == 0) {
        *kind = YAI_RECORD_CARRIER_REQUEST;
    } else if (strcmp(value, "effect_receipt") == 0) {
        *kind = YAI_RECORD_EFFECT_RECEIPT;
    } else if (strcmp(value, "filesystem_receipt") == 0) {
        *kind = YAI_RECORD_FILESYSTEM_RECEIPT;
    } else if (strcmp(value, "subject_state") == 0) {
        *kind = YAI_RECORD_SUBJECT_STATE;
    } else if (strcmp(value, "graph_edge") == 0) {
        *kind = YAI_RECORD_GRAPH_EDGE;
    } else if (strcmp(value, "reconstruction") == 0) {
        *kind = YAI_RECORD_RECONSTRUCTION;
    } else if (strcmp(value, "memory_candidate") == 0) {
        *kind = YAI_RECORD_MEMORY_CANDIDATE;
    } else if (strcmp(value, "divergence") == 0) {
        *kind = YAI_RECORD_DIVERGENCE;
    } else if (strcmp(value, "reconciliation") == 0) {
        *kind = YAI_RECORD_RECONCILIATION;
    } else if (strcmp(value, "projection_request") == 0) {
        *kind = YAI_RECORD_PROJECTION_REQUEST;
    } else if (strcmp(value, "projection_result") == 0) {
        *kind = YAI_RECORD_PROJECTION_RESULT;
    } else if (strcmp(value, "query_result") == 0) {
        *kind = YAI_RECORD_QUERY_RESULT;
    } else if (strcmp(value, "projection_rule") == 0) {
        *kind = YAI_RECORD_PROJECTION_RULE;
    } else if (strcmp(value, "authority_scope") == 0) {
        *kind = YAI_RECORD_AUTHORITY_SCOPE;
    } else if (strcmp(value, "model_interpretation") == 0) {
        *kind = YAI_RECORD_MODEL_INTERPRETATION;
    } else if (strcmp(value, "case_domain") == 0) {
        *kind = YAI_RECORD_CASE_DOMAIN;
    } else if (strcmp(value, "case_attachment") == 0) {
        *kind = YAI_RECORD_CASE_ATTACHMENT;
    } else if (strcmp(value, "case_binding") == 0) {
        *kind = YAI_RECORD_CASE_BINDING;
    } else if (strcmp(value, "interaction_thread") == 0) {
        *kind = YAI_RECORD_INTERACTION_THREAD;
    } else if (strcmp(value, "interaction_turn") == 0) {
        *kind = YAI_RECORD_INTERACTION_TURN;
    } else if (strcmp(value, "participant_view_frame") == 0) {
        *kind = YAI_RECORD_PARTICIPANT_VIEW_FRAME;
    } else {
        return YAI_ERR_INVALID;
    }

    return YAI_OK;
}

File: engine/yai-engine/src/record.rs
//! YAI - record model
//!
//! Purpose:
//!   Define normalized operational records and their JSON encodings.
//!
//! Ownership:
//!   Record kind taxonomy and record envelope serialization used by the Rust
//!   engine.
//!
//! Boundary:
//!   Does not own LMDB storage, journal file IO or graph semantics.
//!
//! Status:
//!   active

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum RecordKind {
    Case,
    SubjectBinding,
    Attempt,
    Decision,
    Receipt,
    Projection,
    PolicyRule,
    GateResult,
    DecisionBasis,
    Obligation,
    ReceiptRequirement,
    CarrierRequest,
    EffectReceipt,
    FilesystemReceipt,
    SubjectState,
    GraphEdge,
    Reconstruction,
    MemoryCandidate,
    Divergence,
    Reconciliation,
    ProjectionRequest,
    ProjectionResult,
    QueryResult,
    ProjectionRule,
    AuthorityScope,
    ModelInterpretation,
    CaseDomain,
    CaseAttachment,
    CaseBinding,
    InteractionThread,
    InteractionTurn,
    ParticipantViewFrame,
    ReviewRequest,
    ReviewDecision,
    ControlPending,
}

impl RecordKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Case => "case",
            Self::SubjectBinding => "subject_binding",
            Self::Attempt => "attempt",
            Self::Decision => "decision",
            Self::Receipt => "receipt",
            Self::Projection => "projection",
            Self::PolicyRule => "policy_rule",
            Self::GateResult => "gate_result",
            Self::DecisionBasis => "decision_basis",
            Self::Obligation => "obligation",
            Self::ReceiptRequirement => "receipt_requirement",
            Self::CarrierRequest => "carrier_request",
            Self::EffectReceipt => "effect_receipt",
            Self::FilesystemReceipt => "filesystem_receipt",
            Self::SubjectState => "subject_state",
            Self::GraphEdge => "graph_edge",
            Self::Reconstruction => "reconstruction",
            Self::MemoryCandidate => "memory_candidate",
            Self::Divergence => "divergence",
            Self::Reconciliation => "reconciliation",
            Self::ProjectionRequest => "projection_request",
            Self::ProjectionResult => "projection_result",
            Self::QueryResult => "query_result",
            Self::ProjectionRule => "projection_rule",
            Self::AuthorityScope => "authority_scope",
            Self::ModelInterpretation => "model_interpretation",
            Self::CaseDomain => "case_domain",
            Self::CaseAttachment => "case_attachment",
            Self::CaseBinding => "case_binding",
            Self::InteractionThread => "interaction_thread",
            Self::InteractionTurn => "interaction_turn",
            Self::ParticipantViewFrame => "participant_view_frame",
            Self::ReviewRequest => "review_request",
            Self::ReviewDecision => "review_decision",
            Self::ControlPending => "control_pending",
        }
    }

    pub fn from_str(value: &str) -> Option<Self> {
        match value {
            "case" => Some(Self::Case),
            "subject_binding" => Some(Self::SubjectBinding),
            "attempt" => Some(Self::Attempt),
            "decision" => Some(Self::Decision),
            "receipt" => Some(Self::Receipt),
            "projection" => Some(Self::Projection),
            "policy_rule" => Some(Self::PolicyRule),
            "gate_result" => Some(Self::GateResult),
            "decision_basis" => Some(Self::DecisionBasis),
            "obligation" => Some(Self::Obligation),
            "receipt_requirement" => Some(Self::ReceiptRequirement),
            "carrier_request" => Some(Self::CarrierRequest),
            "effect_receipt" => Some(Self::EffectReceipt),
            "filesystem_receipt" => Some(Self::FilesystemReceipt),
            "subject_state" => Some(Self::SubjectState),
            "graph_edge" => Some(Self::GraphEdge),
            "reconstruction" => Some(Self::Reconstruction),
            "memory_candidate" => Some(Self::MemoryCandidate),
            "divergence" => Some(Self::Divergence),
            "reconciliation" => Some(Self::Reconciliation),
            "projection_request" => Some(Self::ProjectionRequest),
            "projection_result" => Some(Self::ProjectionResult),
            "query_result" => Some(Self::QueryResult),
            "projection_rule" => Some(Self::ProjectionRule),
            "authority_scope" => Some(Self::AuthorityScope),
            "model_interpretation" => Some(Self::ModelInterpretation),
            "case_domain" => Some(Self::CaseDomain),
            "case_attachment" => Some(Self::CaseAttachment),
            "case_binding" => Some(Self::CaseBinding),
            "interaction_thread" => Some(Self::InteractionThread),
            "interaction_turn" => Some(Self::InteractionTurn),
            "participant_view_frame" => Some(Self::ParticipantViewFrame),
            "review_request" => Some(Self::ReviewRequest),
            "review_decision" => Some(Self::ReviewDecision),
            "control_pending" => Some(Self::ControlPending),
            _ => None,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Record {
    pub id: String,
    pub case_ref: String,
    pub kind: RecordKind,
    pub subject_ref: String,
    pub attempt_id: String,
    pub decision_id: String,
    pub receipt_id: String,
    pub summary: String,
}

impl Record {
    pub fn new(id: impl Into<String>, kind: RecordKind, summary: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            case_ref: String::new(),
            kind,
            subject_ref: String::new(),
            attempt_id: String::new(),
            decision_id: String::new(),
            receipt_id: String::new(),
            summary: summary.into(),
        }
    }

    pub fn from_parts(
        id: impl Into<String>,
        case_ref: impl Into<String>,
        kind: RecordKind,
        subject_ref: impl Into<String>,
        attempt_id: impl Into<String>,
        decision_id: impl Into<String>,
        receipt_id: impl Into<String>,
        summary: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            case_ref: case_ref.into(),
            kind,
            subject_ref: subject_ref.into(),
            attempt_id: attempt_id.into(),
            decision_id: decision_id.into(),
            receipt_id: receipt_id.into(),
            summary: summary.into(),
        }
    }

    pub fn to_jsonl(&self) -> String {
        format!(
            "{{\"schema\":\"yai.store.record.v0\",\"record_id\":\"{}\",\"case_ref\":\"{}\",\"record_kind\":\"{}\",\"subject_ref\":\"{}\",\"attempt_id\":\"{}\",\"decision_id\":\"{}\",\"receipt_id\":\"{}\",\"summary\":\"{}\"}}\n",
            escape_json(&self.id),
            escape_json(&self.case_ref),
            self.kind.as_str(),
            escape_json(&self.subject_ref),
            escape_json(&self.attempt_id),
            escape_json(&self.decision_id),
            escape_json(&self.receipt_id),
            escape_json(&self.summary)
        )
    }

    pub fn to_record_plane_json(&self, source_ref: &str) -> String {
        let subject_refs = if self.subject_ref.is_empty() || self.subject_ref == "subject:none" {
            String::new()
        } else {
            format!("\"{}\"", escape_json(&self.subject_ref))
        };
        format!(
            "{{\"schema\":\"yai.record.v1\",\"record_id\":\"{}\",\"record_kind\":\"{}\",\"case_ref\":\"{}\",\"subject_refs\":[{}],\"created_at_unix_ms\":0,\"source\":{{\"plane\":\"journal\",\"ref\":\"{}\"}},\"provenance_refs\":[],\"payload\":{{\"summary\":\"{}\",\"subject_ref\":\"{}\",\"attempt_id\":\"{}\",\"decision_id\":\"{}\",\"receipt_id\":\"{}\"}}}}",
            escape_json(&self.id),
            self.kind.as_str(),
            escape_json(&self.case_ref),
            subject_refs,
            escape_json(source_ref),
            escape_json(&self.summary),
            escape_json(&self.subject_ref),
            escape_json(&self.attempt_id),
            escape_json(&self.decision_id),
            escape_json(&self.receipt_id)
        )
    }

    pub fn from_jsonl(line: &str) -> Option<Self> {
        let schema = extract_json_string(line, "schema")?;
        if schema != "yai.store.record.v0" {
            return None;
        }
        let kind = RecordKind::from_str(&extract_json_string(line, "record_kind")?)?;
        Some(Self::from_parts(
            extract_json_string(line, "record_id")?,
            extract_json_string(line, "case_ref").unwrap_or_default(),
            kind,
            extract_json_string(line, "subject_ref").unwrap_or_default(),
            extract_json_string(line, "attempt_id").unwrap_or_default(),
            extract_json_string(line, "decision_id").unwrap_or_default(),
            extract_json_string(line, "receipt_id").unwrap_or_default(),
            extract_json_string(line, "summary").unwrap_or_default(),
        ))
    }
}

fn escape_json(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
        .replace('\r', "\\r")
}

fn extract_json_string(line: &str, key: &str) -> Option<String> {
    let pattern = format!("\"{key}\":\"");
    let start = line.find(&pattern)? + pattern.len();
    let mut result = String::new();
    let mut chars = line[start..].chars();
    while let Some(ch) = chars.next() {
        match ch {
            '"' => return Some(result),
            '\\' => {
                let escaped = chars.next()?;
                match escaped {
                    'n' => result.push('\n'),
                    'r' => result.push('\r'),
                    other => result.push(other),
                }
            }
            other => result.push(other),
        }
    }
    None
}
